src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs

```diff
@@ -1759,6 +1763,12 @@ private Symbol GetDeclaredMember(NamespaceOrTypeSymbol container, TextSpan decla
 
             // look for any member with same declaration location
             var collection = name != null ? container.GetMembers(name) : container.GetMembersUnordered();
+            if (isKnownNamespace && name != null && collection is [NamespaceSymbol knownNamespace])
```

Comment:
>collection is [NamespaceSymbol knownNamespace] [](http://example.com/codeflow?start=52&length=46)

Let's restructure the code so that we could assert that this condition is true when ```isKnownNamespace``` is true. We should still check the condition for real.

Reply:
I'm concerned that the condition would not hold for certain real-world code, for example:

```
namespace Outer {
  namespace Inner { }
  class Inner { }
}
```

Reply:
➡️ Added this assertion. Will review what to do if tests fail.

Reply:
> I'm concerned that the condition would not hold for certain real-world code, for example

I think it would be good to be proactive and add an explicit test for this scenario. Quite possibly we would like to optimize scenarios like that too. Rather than checking for ```collection is [NamespaceSymbol knownNamespace]```, probably we should use ```collection.OfType<NamespaceSymbol>().AsSingleton() is NamespaceSymbol knownNamespace```. And assert that instead too

Reply:
➡️ Updated to use `AsSingleton`, but in a slightly different non-allocating form.